### PR TITLE
PR: Add testing for comparison of original and generated images with extension

### PR DESCRIPTION
**Pull request by goanpeca**
This is ready! 💥

### CODE REVIEW

**Comment by saulshanabrook** [comment]
@goanpeca What do you think about uploading the artifacts (screenshots or SVGs, whatever you are comparing) after the tests completes so we can look at them as well? https://github.com/jupyterlab/jupyterlab-data-explorer/blob/913375c18a0b195159166adeb927758a46b7b1ca/.github/workflows/test.yml#L37-L42

**Comment by goanpeca** [comment]
>@goanpeca What do you think about uploading the artifacts (screenshots or SVGs, whatever you are comparing) after the tests completes so we can look at them as well?

Yes! next on my todo list

**Comment by goanpeca** [comment]
@saulshanabrook I think this one is ready.

I added the benchmark images to the repo and used [ResembleJS](https://github.com/rsmbl/Resemble.js) to run the compare with a 1% threshold (to evaluate later 🤷‍♂ ?)

The test run on `headless: false` mode and this is a small gif of how it looks locally:
example


The test prints a bunch os stuff that looks like:
```
2019-12-13T01:45:37.4462951Z > env JEST_PUPPETEER_CONFIG=./config/jest/jest-puppeteer.config.js jest --runInBand --verbose=true --detectOpenHandles --config ./config/jest/jest.config.js
2019-12-13T01:45:37.4463340Z 
2019-12-13T01:45:43.5912798Z [I 01:45:43.590 LabApp] Writing notebook server cookie secret to /home/runner/.local/share/jupyter/runtime/notebook_cookie_secret
2019-12-13T01:45:43.8178461Z [W 01:45:43.817 LabApp] All authentication is disabled.  Anyone who can connect to this server will be able to run code.
2019-12-13T01:45:50.2717487Z [I 01:45:45.750 LabApp] JupyterLab extension loaded from /usr/share/miniconda/envs/ibis-vega-transform/lib/python3.7/site-packages/jupyterlab
2019-12-13T01:45:50.2722313Z [I 01:45:45.750 LabApp] JupyterLab application directory is /usr/share/miniconda/envs/ibis-vega-transform/share/jupyter/lab
2019-12-13T01:45:50.2722529Z [W 01:45:47.198 LabApp] JupyterLab server extension not enabled, manually loading...
2019-12-13T01:45:50.2722988Z [I 01:45:48.885 LabApp] JupyterLab extension loaded from /usr/share/miniconda/envs/ibis-vega-transform/lib/python3.7/site-packages/jupyterlab
2019-12-13T01:45:50.2723398Z [I 01:45:48.886 LabApp] JupyterLab application directory is /usr/share/miniconda/envs/ibis-vega-transform/share/jupyter/lab
2019-12-13T01:45:50.3674842Z   console.log test/ui/test.ts:229
2019-12-13T01:45:50.3675382Z     
2019-12-13T01:45:50.3675651Z     
2019-12-13T01:45:50.3676438Z     ## Running "charting-example.ipynb"
2019-12-13T01:45:50.3676767Z     
2019-12-13T01:45:50.3676985Z 
2019-12-13T01:45:50.5279183Z [I 01:45:50.527 LabApp] Serving notebooks from local directory: /home/runner/work/ibis-vega-transform/ibis-vega-transform
2019-12-13T01:45:50.5286530Z [I 01:45:50.527 LabApp] The Jupyter Notebook is running at:
2019-12-13T01:45:50.5286996Z [I 01:45:50.527 LabApp] http://localhost:8080/
2019-12-13T01:45:50.5287662Z [I 01:45:50.527 LabApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
2019-12-13T01:45:53.9425915Z [I 01:45:53.938 LabApp] Writing notebook-signing key to /home/runner/.local/share/jupyter/notebook_secret
2019-12-13T01:45:53.9431112Z [W 01:45:53.939 LabApp] Notebook examples/charting-example.ipynb is not trusted
2019-12-13T01:45:54.5432189Z [I 01:45:54.542 LabApp] Kernel started: 699484ae-c4c0-4f7e-800b-63cb1d19ca78
2019-12-13T01:46:03.0044425Z [I 01:46:03.001 LabApp] Kernel restarted: 699484ae-c4c0-4f7e-800b-63cb1d19ca78
2019-12-13T01:46:07.0585521Z [I 01:46:07.056 LabApp] Build is up to date
2019-12-13T01:46:12.1119784Z INFO:     Started server process [5834]
2019-12-13T01:46:12.1127419Z INFO:     Uvicorn running on http://0.0.0.0:56045 (Press CTRL+C to quit)
2019-12-13T01:46:12.1131713Z INFO:     Waiting for application startup.
2019-12-13T01:46:12.1132068Z INFO:     Application startup complete.
2019-12-13T01:46:12.4327007Z INFO:     Reporting span a3f57791b5e4d24a:5851fbdb79c3e514:45416e53b6fa7637:1 browser.compileSpec
2019-12-13T01:46:12.7157796Z INFO:     Reporting span a3f57791b5e4d24a:694245c56e0e59a5:225361924e6679e5:1 browser.comm:queryibis
2019-12-13T01:46:12.7279570Z INFO:     Reporting span a3f57791b5e4d24a:225361924e6679e5:282a39825629f13b:1 browser.transform
2019-12-13T01:46:13.0111482Z INFO:     Reporting span a3f57791b5e4d24a:ba01518c33657dac:f2d0f10c26896cef:1 browser.comm:queryibis
2019-12-13T01:46:13.0216547Z INFO:     Reporting span a3f57791b5e4d24a:f2d0f10c26896cef:282a39825629f13b:1 browser.transform
2019-12-13T01:46:13.2425243Z INFO:     Reporting span a3f57791b5e4d24a:38a939b8d82a9dd6:dfa14928b2d9f054:1 browser.comm:queryibis
2019-12-13T01:46:13.2494916Z INFO:     Reporting span a3f57791b5e4d24a:dfa14928b2d9f054:282a39825629f13b:1 browser.transform
2019-12-13T01:46:13.4851964Z INFO:     Reporting span a3f57791b5e4d24a:19e24de5fbff7815:b637e16e902eb15:1 browser.comm:queryibis
2019-12-13T01:46:13.4985994Z INFO:     Reporting span a3f57791b5e4d24a:b637e16e902eb15:282a39825629f13b:1 browser.transform
2019-12-13T01:46:13.7555153Z INFO:     Reporting span a3f57791b5e4d24a:5662bfdc842b8c2c:12985883fe42adb1:1 browser.comm:queryibis
2019-12-13T01:46:13.7644716Z INFO:     Reporting span a3f57791b5e4d24a:12985883fe42adb1:282a39825629f13b:1 browser.transform
2019-12-13T01:46:14.0130133Z INFO:     Reporting span a3f57791b5e4d24a:c413d41f87e7084:a2ce3b1c27304a4a:1 browser.comm:queryibis
2019-12-13T01:46:14.0240349Z INFO:     Reporting span a3f57791b5e4d24a:a2ce3b1c27304a4a:282a39825629f13b:1 browser.transform
2019-12-13T01:46:14.3317299Z INFO:     Reporting span a3f57791b5e4d24a:e329b2271d1ca891:4e078f285a469c94:1 browser.comm:queryibis
2019-12-13T01:46:14.3432929Z INFO:     Reporting span a3f57791b5e4d24a:4e078f285a469c94:282a39825629f13b:1 browser.transform
2019-12-13T01:46:14.6381998Z INFO:     Reporting span a3f57791b5e4d24a:e0758148e30c20a0:5c5b42cf65a09f67:1 browser.comm:queryibis
2019-12-13T01:46:14.6519301Z INFO:     Reporting span a3f57791b5e4d24a:5c5b42cf65a09f67:282a39825629f13b:1 browser.transform
2019-12-13T01:46:14.7589894Z INFO:     Reporting span a3f57791b5e4d24a:1cf88b955b690ea3:45416e53b6fa7637:1 browser.vegaEmbed
2019-12-13T01:46:14.8034228Z INFO:     Reporting span a3f57791b5e4d24a:45416e53b6fa7637:282a39825629f13b:1 browser.renderModel
2019-12-13T01:46:14.9245171Z   console.log test/ui/test.ts:253
2019-12-13T01:46:14.9245616Z     1
2019-12-13T01:46:14.9245746Z 
2019-12-13T01:46:16.9379760Z   console.log test/ui/test.ts:261
2019-12-13T01:46:16.9381325Z     Setting download folder: /home/runner/work/ibis-vega-transform/images/charting-example
2019-12-13T01:46:16.9381722Z 
2019-12-13T01:46:21.9562179Z   console.log test/ui/test.ts:164
2019-12-13T01:46:21.9563334Z     File saved at: /home/runner/work/ibis-vega-transform/images/charting-example/visualization.svg
2019-12-13T01:46:21.9563669Z 
2019-12-13T01:46:21.9563951Z   console.log test/ui/test.ts:165
2019-12-13T01:46:21.9564760Z     File renamed to: /home/runner/work/ibis-vega-transform/images/charting-example/0.svg
2019-12-13T01:46:26.3871943Z   console.log test/ui/test.ts:277
2019-12-13T01:46:26.3873341Z     Original images folder: /home/runner/work/ibis-vega-transform/ibis-vega-transform/images/charting-example
2019-12-13T01:46:26.3873628Z 
2019-12-13T01:46:26.3873882Z   console.log test/ui/test.ts:199
2019-12-13T01:46:26.3874467Z     Comparing images: "0.svg"
2019-12-13T01:46:26.3874735Z 
2019-12-13T01:46:26.3874968Z   console.log test/ui/test.ts:200
2019-12-13T01:46:26.3875631Z     /home/runner/work/ibis-vega-transform/ibis-vega-transform/images/charting-example/0.svg
2019-12-13T01:46:26.3875863Z 
2019-12-13T01:46:26.3876089Z   console.log test/ui/test.ts:201
2019-12-13T01:46:26.3876498Z     /home/runner/work/ibis-vega-transform/images/charting-example/0.svg
2019-12-13T01:46:26.3876716Z 
2019-12-13T01:46:26.3876939Z   console.log test/ui/test.ts:209
2019-12-13T01:46:26.3877338Z     {
2019-12-13T01:46:26.3877824Z       isSameDimensions: true,
2019-12-13T01:46:26.3880240Z       dimensionDifference: { width: 0, height: 0 },
2019-12-13T01:46:26.3880422Z       rawMisMatchPercentage: 0,
2019-12-13T01:46:26.3880821Z       misMatchPercentage: '0.00',
2019-12-13T01:46:26.3880934Z       diffBounds: { top: 991, left: 1119, bottom: 0, right: 0 },
2019-12-13T01:46:26.3881041Z       analysisTime: 2157,
2019-12-13T01:46:26.3881136Z       getImageDataUrl: [Function]
2019-12-13T01:46:26.3881587Z     }
2019-12-13T01:46:26.3881620Z 
2019-12-13T01:46:31.9625720Z [I 01:46:31.961 LabApp] Kernel shutdown: 699484ae-c4c0-4f7e-800b-63cb1d19ca78
2019-12-13T01:46:35.7520182Z   console.log test/ui/test.ts:229
```

File: README.md [hunk 1]
```diff
@@ -56,20 +56,19 @@ If you want to see traces of the interactiosn for debugging and performance anal
 install the `jaeger-all-in-one` binary and the `jupyterlab-server-proxy-saulshanabrook`
 lab extension to see the Jaeger icon in the launcher.
 
-
 ```bash
-conda install -c conda-forge
+conda install jaeger -c conda-forge
```
Comment (goanpeca) [inline README.md]: I think on the latest PR the `jaeger` was missing @saulshanabrook

File: binder/environment.yml [hunk 2]
```diff
@@ -2,10 +2,10 @@ name: ibis-vega-transform
...
+  - jupyterlab
   - pymapd
-  - nodejs
   - python>=3.7
-  - jupyterlab
+  - nodejs=12
   - notebook<6.0
-  - jaeger
-
+  - tornado<6
```
Comment (goanpeca) [inline binder/environment.yml]: Avoids changing tornado when running pip install

File: config/jest/jest-puppeteer.config.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ * @license BSD-3-Clause
+ *
+ * Copyright (c) 2019 Project Jupyter Contributors.
+ * Distributed under the terms of the 3-Clause BSD License.
+ */
+
+const portNumber = 8080;
+const config = {
+  launch: {
+    headless: false,
```
Comment (goanpeca) [inline config/jest/jest-puppeteer.config.js]: Tests require and actual visible browser

File: tsconfig.test.json [hunk 4]
```diff
@@ -0,0 +1,10 @@
+{
+  "extends": "./tsconfig.json",
+  "compilerOptions": {
+    "outDir": "build/test",
+    "rootDir": "test",
+    "strict": false,
+    "noImplicitAny": false
```
Comment (goanpeca) [inline tsconfig.test.json]: This was failing with the tests and I could not debug some errors. Might fix later.

File: README.md (hunk 1 above)
Comment (saulshanabrook) [reply README.md]: Ah thanks!

File: binder/environment.yml (hunk 2 above)
Comment (saulshanabrook) [reply binder/environment.yml]: Great.

**Pull request merged**